### PR TITLE
feat(docs) Adding some additional clarity around low power states

### PR DESCRIPTION
**Pull request by Nick-Munnich**
Minor improvements to resolve some questions that appeared on Discord. Also made a note of the soft-off/deep-sleep wakeup-sources issue, since I couldn't find that documented anywhere.

### CODE REVIEW

File: docs/docs/config/power.md
Comment (caksoylar) [inline docs/docs/config/power.md:15]:
```suggestion
In the deep sleep state, the keyboard additionally disconnects from Bluetooth and any external power output is disabled. This state uses very little power, but it may take a few seconds to reconnect after waking. To allow a device to wake from deep sleep, the `wakeup-source` property needs to be added to the [keyboard scan driver](kscan.md) of said device.
```
The linked page documents the property and has examples as well, so I wouldn't distract with the soft off page.

File: docs/docs/features/soft-off.md
Comment (caksoylar) [inline docs/docs/features/soft-off.md:24]: This sounds alright to me, but to ensure technical correctness I'd appreciate if e.g. @petejohanson gave an OK.

**Review by caksoylar — COMMENTED** [review]
LGTM in general: While touching the soft-off page, do you mind adding a commit to remove `label = ..` properties from the examples in that page? That's the only spot they are left in the docs.

File: docs/docs/features/soft-off.md
Comment (Nick-Munnich) [reply docs/docs/features/soft-off.md:24]: Removed LABEL= as requested. Mind you I have no idea what those do or why they were there, just following instructions. Agree on confirmation of correctness needed.

Comment (caksoylar) [reply docs/docs/features/soft-off.md:24]: Thanks, they used to be needed but they are no longer used as of #2028.

Comment (Nick-Munnich) [reply docs/docs/features/soft-off.md:24]: Reading through this again on a different day, I have some unresolved questions on how it behaves and would like for them to be confirmed so that I can adjust the text to be clearer:
- Can you define multiple GPIO pins for soft-off? (Pretty sure yes)
- If multiple are defined, do you need to trigger the *same* soft-off pin to wake up, or can you trigger any? (Pretty sure any)
- Can a GPIO be defined for wakeup but not for activating soft-off (i.e. you could use &soft-off to put it to sleep, and press the index finger home to wakeup, I think this one is a no)

**Comment by nVitius** [comment]
Upgraded the firmware on my keyboard recently. Ran into an issue where I couldn't wake it from sleep and couldn't figure out why until I saw this PR.

I'm not sure about the soft-off portion of this PR, but I think the comment about `wakeup-source` on the deep sleep docs is important to have.

File: docs/docs/features/soft-off.md
Comment (petejohanson) [reply docs/docs/features/soft-off.md:24]: > Reading through this again on a different day, I have some unresolved questions on how it behaves and would like for them to be confirmed so that I can adjust the text to be clearer:
> 
>     * Can you define multiple GPIO pins for soft-off? (Pretty sure yes)

I don't see why you couldn't... Not sure the value in doing so, but it should work.

> 
>     * If multiple are defined, do you need to trigger the _same_ soft-off pin to wake up, or can you trigger any? (Pretty sure any)

Any should do it... Under the hood, the only requirements for the *wake* is to generate any of the valid events that wake nRF52 from "System OFF": 
image

> 
>     * Can a GPIO be defined for wakeup but not for activating soft-off (i.e. you could use &soft-off to put it to sleep, and press the index finger home to wakeup, I think this one is a no)

I mean... in theory yes, if set up properly in the DTS. The two things are complementary, not strictly tied together.

**Comment by Nick-Munnich** [comment]
Thanks again Pete for the reply. I have done some larger changes to accommodate. I think this version reads much nicer. Unfortunately a rename was involved which makes reviewing it more difficult. 

Essentially, the soft off feature page is renamed to low power states, and the information about idle and deep sleep were moved into it. I also documented how to reuse a key for wakeup-only, and how to make the &soft_off behave like &sleep. Some minor renaming was involved to make things read nicer.

All the content for the direct pin approach is tested and working. The content for the matrix-integrated and switch matrix wakeup-only was working, but currently isn't: see #2376. It may need some changing after said issue is resolved.

File: docs/docs/features/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states.mdx:1]: I think in general it might be worth removing the root node enclosures (`/ {` and `};` lines) from all the code samples here. It'd shorten the text file a bit as well.

File: docs/docs/features/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states.mdx:298]:
```suggestion
For the wakeup-only approach, all `soft_off` child nodes are omitted.
```

File: docs/docs/features/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states.mdx:312]:
```suggestion
#### Soft off waker
```

File: docs/docs/features/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states.mdx:332]: As per our discussion on Discord, this is necessary with the "wakeup-only" too, right? This below part might be read as it applies only to the combination case:

> This is only needed for matrix integrated combos. For those keyboards, the list should include the matrix output needs needed so the combo hardware is properly "driven" when the keyboard is off.

Maybe this bullet point just needs to be simplified, given we are already in the "matrix integrated" tab?

File: docs/docs/features/low-power-states.mdx
Comment (caksoylar) [reply docs/docs/features/low-power-states.mdx:135]: Sounds good.

File: docs/docs/features/low-power-states.mdx
Comment (caksoylar) [reply docs/docs/features/low-power-states.mdx:88]: 👍

File: docs/docs/features/low-power-states.mdx
Comment (caksoylar) [reply docs/docs/features/low-power-states.mdx]: Sure, I think we can cross that bridge when we get to it.

File: docs/docs/features/low-power-states.mdx
Comment (caksoylar) [reply docs/docs/features/low-power-states.mdx:61]: I counted an additional soft off switch (like a push button wired to direct pin) as additional hardware as well. But if that doesn't count, I think "combination" is still better than "combo".

File: docs/docs/features/low-power-states.mdx
Comment (caksoylar) [reply docs/docs/features/low-power-states.mdx:1]: OK, though that's definitely not something consistent in the docs. If you want to keep them, let's make it consistent across this page so all snippets have it (including the space between `/ {`).

File: docs/docs/features/low-power-states.mdx
Comment (caksoylar) [reply docs/docs/features/low-power-states.mdx:287]: From a cursory look at `src/kscan_sideband_behaviors.c` it looks like both are checked in the callback invoked by the kscan driver, so I guess it is important which output pin generates the event as well.

File: docs/docs/features/low-power-states.mdx
Comment (Nick-Munnich) [reply docs/docs/features/low-power-states.mdx]: I've rephrased this as part of reshuffling to make the wakeup-only approach clearer.

File: docs/docs/features/low-power-states.mdx
Comment (Nick-Munnich) [reply docs/docs/features/low-power-states.mdx]: Simplified as part of reshuffling

**Comment by Nick-Munnich** [comment]
Did some reshuffling to try and get the tabs behaving nicely without too much duplication. Since this required shoving a bunch of content into other files, I also did so for some content that doesn't get duplicated because it makes the file overall much easier to read.

The wakeup-only approach is much simplified and should be clearer now, with minor changes to the direct and matrix approaches as a result of the reshuffling.

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states/low-power-states.mdx:67]:
```suggestion
- Triggering the [soft off behavior](../../behaviors/soft-off.md) from the keymap.
```

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states/low-power-states.mdx:84]:
```suggestion
For keyboards which do not have such hardware, using soft off is as simple as placing the [soft off behavior](../../behaviors/soft-off.md) in your keymap and then invoking it.
```

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states/low-power-states.mdx:74]:
```suggestion
The GPIO pin used to wake from sleep can be a hardware-defined one, such as for a dedicated on-off push button, or it can be a single specific key switch reused for waking up (which may be accidentally pressed, e.g. while the device is being carried in a bag). To allow the simultaneous pressing of multiple keyswitches to trigger and exit soft off, some keyboards make use of additional hardware to integrate the dedicated GPIO pin into the keyboard matrix.
```

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states/low-power-states.mdx:105]:
```suggestion

### Hardware Changes
```

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states/low-power-states.mdx:1]: If we rename to `index.mdx` can we use `features/low-power-states` in the sidebar?

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states/low-power-states.mdx:251]: Should we put this under "wakeup-only" tab? It is essentially wakeup-only but with any key, isn't it?

File: docs/docs/features/low-power-states/_sideband-wakeup-matrix.md
Comment (caksoylar) [inline docs/docs/features/low-power-states/_sideband-wakeup-matrix.md:8]: Double checking: does this not need `auto-enable`?

File: docs/docs/features/low-power-states/_sideband-matrix.md
Comment (caksoylar) [inline docs/docs/features/low-power-states/_sideband-matrix.md:48]:
```suggestion
```
I suppose this shouldn't be here anymore

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states/low-power-states.mdx:138]:
```suggestion

#### Soft off behavior
```

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states/low-power-states.mdx:225]:
```suggestion
    <SoftOffWaker />
```

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (caksoylar) [inline docs/docs/features/low-power-states/low-power-states.mdx:213]:
```suggestion
    You also need to update the `zmk,kscan` chosen value to point to the new kscan instance:
```

**Review by caksoylar — COMMENTED** [review]
The wakeup-only case is much clearer for me with these changes, thanks.

Couple general replacements:
- [Bb]"ehaviour" -> "ehavior" (including filenames?)
- [Kk]"eyswitch" -> "ey switch"

It would also be good to add `dts` to the fenced code blocks, since we get syntax highlighting for those.

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (Nick-Munnich) [reply docs/docs/features/low-power-states/low-power-states.mdx:253]: We certainly could and it would make sense. I quite like it where it is because it gives a bit better intuition of what `wakeup-sources` does, and I don't think it's causing any confusion at that location.

File: docs/docs/features/low-power-states/_sideband-wakeup-matrix.md
Comment (Nick-Munnich) [reply docs/docs/features/low-power-states/_sideband-wakeup-matrix.md:8]: Neither approach needs `auto-enable` prior to [048cb0](https://github.com/zmkfirmware/zmk/commit/0438cb0ee532a31291aee3dabe24362f16146fd2). Since matrix integrated solutions have been broken since then (#2376), I'm not sure how to answer that. Pete's changes in said commit didn't add `auto-enable` to the docs for the matrix case, so I haven't included it either.

File: docs/docs/features/low-power-states/low-power-states.mdx
Comment (Nick-Munnich) [reply docs/docs/features/low-power-states/low-power-states.mdx:1]: Unfortunately it doesn't seem like it, just did a simple test.

**Comment by Nick-Munnich** [comment]
Made the dts and general replacements.

File: docs/docs/features/low-power-states/_sideband-wakeup-matrix.md
Comment (caksoylar) [reply docs/docs/features/low-power-states/_sideband-wakeup-matrix.md:8]: I guess we'll ask Pete about it when he can get around to it.

**Review by caksoylar — APPROVED** [review]
I am happy with this, thanks again for tackling it. However I'd appreciate if @petejohanson could go over [the finished page](https://deploy-preview-2330--zmk.netlify.app/docs/features/low-power-states) to confirm whether
- there are any technical correctness issues I missed
- he has any concerns on documenting the wakeup-only approach

**Comment by Nick-Munnich** [comment]
Did a refactoring of this to make it cleaner (as was discussed previously on Discord). Originally I intended to do the refactor after this is merged, but as it has been slow-going I decided to put in the effort to separate out the content. Should make everything easier to review as well.

I added what information I could to the config section, but it may need double-checking.

**Comment by Nick-Munnich** [comment]
Folded the content from #2452 back into this - trying to get it nicely into separate PRs post development reshuffling was annoying me too much.

**Comment by petejohanson** [comment]
> I am happy with this, thanks again for tackling it. However I'd appreciate if @petejohanson could go over [the finished page](https://deploy-preview-2330--zmk.netlify.app/docs/features/low-power-states) to confirm whether
> 
>     * there are any technical correctness issues I missed
> 
>     * he has any concerns on documenting the wakeup-only approach

All looks reasonable to me. The description of deep sleep is a tad... non-technical, and that may be the intention. If so, looks fine to me. If we want to offer more details on the specifics of the power state of the MCU, RAM retention (or not), etc., I'm not sure if that's best served living on that page or elsewhere.

**Comment by Nick-Munnich** [comment]
> If we want to offer more details on the specifics of the power state of the MCU, RAM retention (or not), etc., I'm not sure if that's best served living on that page or elsewhere.

This is a good idea. I'll have a go at adding such specifics when I touch this page up a bit again to get it ready for merging.

File: docs/docs/features/low-power-states.md
Comment (caksoylar) [inline docs/docs/features/low-power-states.md:41]: This ideally wouldn't be here, since it is more hardware integration than user-facing. Maybe we can link to the new-shield kscan section, and say "your keyboard should be configured with this by default"?

File: docs/docs/keymaps/behaviors/soft-off.md
Comment (caksoylar) [inline docs/docs/keymaps/behaviors/soft-off.md:1]: This page https://deploy-preview-2330--zmk.netlify.app/docs/development/hardware-integration/soft-off-setup has some issues with duplicate items in the sidebar, maybe due to the tabs? 
image

File: docs/docs/features/low-power-states.md
Comment (Nick-Munnich) [reply docs/docs/features/low-power-states.md:41]: I see your point, but I'd like to leave it here for now. It's aimed towards those whose keyboards are custom and haven't been updated with the property yet. I'm happy to remove it and rephrase once we're reasonably confident that almost everyone affected has modified their keyboard. My estimate would be a few months once we have versioning, Zephyr 3.7, and a few other things.

File: docs/docs/keymaps/behaviors/soft-off.md
Comment (Nick-Munnich) [reply docs/docs/keymaps/behaviors/soft-off.md:1]: Yeah, tabs are the cause. I just changed the header depth of the page, reshuffling to try and get rid of this would get ugly I think.

File: docs/docs/features/low-power-states.md
Comment (caksoylar) [reply docs/docs/features/low-power-states.md:41]: It's been 6+ months and I haven't been seeing a lot of issues related to the missing property lately, but we can leave it if you want to. I hope it doesn't cause more confusion than it solves.

File: docs/docs/development/hardware-integration/soft-off-setup.mdx
Comment (caksoylar) [inline docs/docs/development/hardware-integration/soft-off-setup.mdx:157]:
```suggestion
If you add your kscan to the `wakeup-sources` array, then your keyboard will wake upon pressing any key in your kscan. Essentially, this causes `&soft_off` to behave like a behavior that puts the keyboard in deep sleep. If you choose to do so, then you can omit everything aside from the `soft_off_wakers` node.
```

I haven't found a reference to `&sleep` in the page so I think it might not be intuitively clear what it would supposedly be for.

File: docs/docs/features/low-power-states.md
Comment (caksoylar) [inline docs/docs/features/low-power-states.md:16]:
```suggestion
- If possible, disables external power output
```

File: docs/docs/features/low-power-states.md
Comment (caksoylar) [inline docs/docs/features/low-power-states.md:51]:
```suggestion
The device enters the same software power-off state as in deep sleep, but is significantly more restrictive in the sources which can wake it. Power is _not_ technically removed from the entire system, unlike a hardware switch.
```

**Review by caksoylar — APPROVED** [review]
Only a few small notes, LGTM otherwise.

**Pull request merged**
